plugins/woocommerce-admin/client/products/product-title.tsx

```diff
@@ -16,42 +18,112 @@ import { useSelect } from '@wordpress/data';
  * Internal dependencies
  */
 import { getProductTitle } from './utils/get-product-title';
+import { getProductVariationTitle } from './utils/get-product-variation-title';
 import { ProductBreadcrumbs } from './product-breadcrumbs';
 import { ProductStatusBadge } from './product-status-badge';
 import { WooHeaderPageTitle } from '~/header/utils';
 import './product-title.scss';
 
 export const ProductTitle: React.FC = () => {
 	const { values } = useFormContext< Product >();
-	const { productId } = useParams();
-	const { persistedName } = useSelect( ( select: WCDataSelector ) => {
-		const product = productId
-			? select( PRODUCTS_STORE_NAME ).getProduct(
+	const { productId, variationId } = useParams();
+	const { isLoading, persistedName, productVariation } = useSelect(
+		( select: WCDataSelector ) => {
+			const {
+				getProduct,
+				hasFinishedResolution: hasProductFinishedResolution,
+			} = select( PRODUCTS_STORE_NAME );
+			const {
+				getProductVariation,
+				hasFinishedResolution: hasProductVariationFinishedResolution,
+			} = select( EXPERIMENTAL_PRODUCT_VARIATIONS_STORE_NAME );
+			const product = productId
+				? getProduct( parseInt( productId, 10 ) )
+				: null;
+			const variation =
+				variationId && productId
+					? getProductVariation( {
+							id: parseInt( variationId, 10 ),
+							product_id: parseInt( productId, 10 ),
+					  } )
+					: null;
+
+			const isProductLoading =
+				productId &&
+				! hasProductFinishedResolution( 'getProduct', [
 					parseInt( productId, 10 ),
-					undefined
-			  )
-			: null;
+				] );
+
+			const isVariationLoading =
+				variationId &&
+				productId &&
+				! hasProductVariationFinishedResolution(
+					'getProductVariation',
+					[
+						{
+							id: parseInt( variationId, 10 ),
+							product_id: parseInt( productId, 10 ),
+						},
+					]
+				);
 
-		return {
-			persistedName: product?.name,
-		};
-	} );
+			return {
+				persistedName: product?.name,
+				productVariation: variation,
+				isLoading: isProductLoading || isVariationLoading,
+			};
+		}
+	);
+
+	const productTitle = getProductTitle(
+		values.name,
+		values.type,
+		persistedName
+	);
+	const productVariationTitle =
+		productVariation && getProductVariationTitle( productVariation );
 
-	const breadcrumbs = [
+	const pageHierarchy = [
```

Comment:
I think it might be beneficial to make use of `useMemo` here instead of re-generating this list on every render (currently that is every form change, hopefully that will be fixed though ).

Reply:
I opted not to use `useMemo` here since this is not an expensive operation (3 item array), but definitely agree that we need to fix the form re-rendering so all components can benefit from the performance fixes.

Happy to include `useMemo` here though if you feel strongly about it.

Reply:
In this case I am not to worried about the `filter` operation as that will indeed hardly cost anything for a 3 item array.
But more so that we are creating a new reference to `pageHierarchy` which in turn will trigger the render method of `ProductBreadcrumbs` component as the prop changes on every render.
Now product breadcrumbs is also relatively lightweight and makes use of the `href` as the key so that shouldn't be to problematic either.
So I am fine fore-going it for now, but in cases like these were we pass an object down to another component, we probably want to be aware if that component does something more expensive with the data.
So in theory it might be good practice for use to make use of `useMemo` more often so we don't have to come back after the fact.

Sorry, this reply end up being a lot longer :|